vmdb/app/models/metric/ci_mixin/capture/openstack_base.rb

```diff
@@ -37,134 +37,355 @@ def perf_release_openstack
+            log_warn_data_corrupted(i, log_header, multi_counter_aligned_start_guard, last_period)
           end
+          multi_counter_metrics      = {}
+          last_multi_counter_metrics = {}
+          # Moving to guard to another period cause the old one had incomplete data for all counters. We will
+          # try to collect all data for all counters in next guard period.
+          multi_counter_aligned_start_guard = last_period
         end
 
-        # For (temporary) symmetry with VIM API we create 20-second intervals.
-        (last_ts + 20.seconds..ts).step_value(20.seconds).each do |ts_item|
-          counter_values_by_ts.store_path(ts_item.iso8601, i[:vim_style_counter_key], value)
+        i[:openstack_counters].each { |c| last_metrics[c] = metrics_by_counter_name.fetch_path(c, last_period) }
+        all_last_metrics_available = process_multi_counter_metrics(i, last_metrics, last_multi_counter_metrics)
+
+        if !all_last_metrics_available
+          # Resetting multi_counter_aligned_end, which will be initialized to period in next period
+          multi_counter_aligned_end = nil
+          # All last_multi_counter data are not available, lets move to another period and try to capture them
+          next
+        else
+          # Move guard when last_period is complete
+          multi_counter_aligned_start_guard = period
+        end
+      end
+
+      if period - multi_counter_aligned_start_guard > data_collecting_period / 2
+        # The data are corrupted or missing. In order to avoid holes in saved data, this period will be filled
+        # by data of next periods.
+        log_warn_data_corrupted(i, log_header, multi_counter_aligned_start_guard, last_period)
+
+        # We haven't found all the data samples of all the counters half of the data_collecting_period away
+        # from the first data sample, just throw away everything and start over, cause we can't find it in this
+        # timeframe. We can keep last_multi_counter_metrics, cause that has been completed.
+        multi_counter_metrics             = {}
+        # Moving to guard to another period cause the old one had incomplete data for all counters. We will try
+        # to collect all data for all counters in next guard period.
+        multi_counter_aligned_start_guard = period
+        # Moving also aligned end, meaning the incomplete stat has been skipped, but will be filled by value of the
+        # next if we are not in the end of collection period.
+        multi_counter_aligned_end         = period
+      end
+
+      # We are ready for fetching multi_counter_metrics
+      all_metrics_available = process_multi_counter_metrics(i, metrics, multi_counter_metrics)
+
+      if all_metrics_available
+        if metric_capture_module.diff_meter? i[:openstack_counters]
+          # I have both multi_counter_metrics and multi_counter_metrics full of data for each counter, I can compute
```

Comment:
ah, should have been 
# I have both multi_counter_metrics and last_multi_counter_metrics full of data